samples/Cordova/Backend/Providers/AuthorizationProvider.cs

```diff
@@ -72,6 +74,66 @@ public sealed class AuthorizationProvider : OpenIdConnectServerProvider {
             context.Validate(application.RedirectUri);
         }
 
+        public override async Task ValidateTokenRequest(ValidateTokenRequestContext context) {
+            // Note: the OpenID Connect server middleware supports authorization code, refresh token, client credentials
+            // and resource owner password credentials grant types but this authorization provider uses a safer policy
+            // rejecting the last two ones. You may consider relaxing it to support the ROPC or client credentials grant types.
+            if (!context.Request.IsAuthorizationCodeGrantType() && !context.Request.IsRefreshTokenGrantType()) {
+                context.Reject(
+                    error: OpenIdConnectConstants.Errors.UnsupportedGrantType,
+                    description: "Only authorization code and refresh token grant types " +
+                                 "are accepted by this authorization server");
+
+                return;
+            }
+
+            // Note: client authentication is not mandatory for non-confidential client applications like mobile apps
+            // (except when using the client credentials grant type) but this authorization server uses a safer policy
+            // that makes client authentication mandatory and returns an error if client_id or client_secret is missing.
+            // You may consider relaxing it to support the resource owner password credentials grant type
+            // with JavaScript or desktop applications, where client credentials cannot be safely stored.
+            // In this case, call context.Skip() to inform the server middleware the client is not trusted.
+            //if (string.IsNullOrEmpty(context.ClientId) || string.IsNullOrEmpty(context.ClientSecret)) {
```

Comment:
Note: consider calling `context.Skip()` immediately if the client_id is missing or empty or add a check to make it mandatory (both options are fine)